File: nodeTOPFLYtechCodec/MqttClient.js
const mqtt = require('mqtt');
var ByteUtils = require("./ByteUtils")
var CryptoTool = require("./CryptoTool")
var ByteUtils = require("./ByteUtils")
var Decoder = require("./Decoder")
var Encoder = require("./Encoder")
var ObdEncoder = require("./ObdEncoder")
var ObdDecoder = require("./ObdDecoder")
var PersonalAssetDecoder = require("./PersonalAssetDecoder")
var PersonalAssetEncoder = require("./PersonalAssetEncoder")
var moment = require("moment")
Decoder.encryptType = CryptoTool.MessageEncryptType.NONE
var curDecoder = PersonalAssetDecoder
var curEncoder = PersonalAssetEncoder


class MqttClient {
    constructor(imei,handleMessageCallback) {
        this.brokerUrl = "mqtt://192.168.1.251:9083";
        this.imei = imei;
        this.client = null;
        this.subTopic = `${this.imei}_S`;
        this.pubTopic = `${this.imei}_R`;
        this.handleMessageCallback = handleMessageCallback || this.defaultHandleMessage.bind(this);
    }

    defaultHandleMessage(message) {
        console.log(`Default handling for message on topic:`, message);
    }

    connect() {
        return new Promise((resolve, reject) => {
            this.client = mqtt.connect(this.brokerUrl);

            this.client.on('connect', () => {
                console.log(`Connected to MQTT Broker at ${this.brokerUrl}`);


                this.client.subscribe(this.subTopic, (err) => {
                    if (!err) {
                        console.log(`Subscribed to subTopic: ${this.subTopic}`);
                        resolve();
                    } else {
                        console.error('Failed to subscribe to subTopic:', err);
                        reject(err);
                    }
                });
            });

            this.client.on('error', (err) => {
                console.error('MQTT Connection Error:', err);
                reject(err);
            });

            this.client.on('message', (receivedTopic, receivedMessage) => {
                this.handleMessage(receivedTopic, receivedMessage);
            });
        });
    }

    publish(message) {
        if (!this.client || !this.client.connected) {
            throw new Error('Not connected to MQTT Broker');
        }
        var sendHexStr = ByteUtils.bytes2HexString(message,0)
        this.client.publish(this.pubTopic, Buffer.from(sendHexStr,"ascii"));
        console.log(`Published message to subTopic ${this.pubTopic}: ${sendHexStr}`);
    }

    handleMessage(subTopic, message) {
        const asciiMessage = message.toString('ascii');
        console.log(`Received message on subTopic ${subTopic}:`, asciiMessage);
        var hexBytesArray = ByteUtils.hexStringToByte(asciiMessage)

        this.handleMessageCallback(hexBytesArray);
    }

    disconnect() {
        if (this.client && this.client.connected) {
            this.client.end(() => {
                console.log('Disconnected from MQTT Broker');
            });
        }
    }
}

module.exports = MqttClient;

File: TopflytechCodec/pythonTopflytechCodec-3.9/MqttClient.py
import paho.mqtt.client as mqtt


class MqttClient:
    def __init__(self,  imei, handle_message_callback=None):
        self.broker_url = "192.168.1.155"
        self.port = 9083
        self.imei = imei
        self.subTopic = "%s_S" % self.imei
        self.pubTopic = "%s_R" % self.imei
        self.client = mqtt.Client()
        self.handle_message_callback = handle_message_callback or self.default_handle_message


        self.client.on_connect = self.on_connect
        self.client.on_message = self.on_message
        self.client.on_disconnect = self.on_disconnect
        self.client.on_error = self.on_error

    def byte2HexString(self,byteArray, index):
        return ''.join('{:02x}'.format(x) for x in byteArray[index:])
    def hexString2Bytes(self,hexStrByte):
        hexStr = hexStrByte.decode('ascii')
        bytesArray = bytes.fromhex(hexStr)
        return bytesArray

    def connect(self):
        try:
            self.client.connect(self.broker_url,self.port)
            self.client.loop_start()
            print("Connected to MQTT Broker at %s" % self.broker_url)
        except Exception as e:
            print("Connection failed: %s" % e)

    def disconnect(self):
        self.client.loop_stop()
        self.client.disconnect()
        print("Disconnected from MQTT Broker")

    def publish(self, message):
        if not self.client.is_connected():
            raise Exception("Not connected to MQTT Broker")

        sendHexStr = self.byte2HexString(message, 0)
        asciiMessage = sendHexStr.encode('ascii')
        self.client.publish(self.pubTopic, asciiMessage)
        print("Published message to subTopic %s: %s" % (self.pubTopic, asciiMessage))

    def on_connect(self, client, userdata, flags, rc):
        print("MQTT Connected with result code %d" % rc)
        client.subscribe(self.subTopic)
        print("Subscribed to subTopic: %s" % self.subTopic)

    def on_message(self, client, userdata, msg):
        try:
            asciiMessage = msg.payload
            print("Received message on subTopic %s: %s" % (msg.topic, asciiMessage))
            hexBytesArray = self.hexString2Bytes(asciiMessage)
            self.handle_message_callback(hexBytesArray)
        except Exception as e:
            print("Connection failed: %s" % e)

    def on_disconnect(self, client, userdata, rc):
        print("Disconnected from MQTT Broker")

    def on_error(self, client, userdata, level, buf):
        print("MQTT Error: %s" % buf)

    def default_handle_message(self, message):
        print("Default handling for message ")









